File: benchmarks/reference/nbody.py
import time
import math

class Body:
    def __init__(self, x, y, z, mass):
        self.x = x
        self.y = y
        self.z = z
        self.mass = mass
        self.vx = 0
        self.vy = 0
        self.vz = 0

def advance(bodies, dt):
    n = len(bodies)
    for i in range(n):
        for j in range(i + 1, n):
            b1 = bodies[i]
            b2 = bodies[j]
            
            dx = b1.x - b2.x
            dy = b1.y - b2.y
            dz = b1.z - b2.z
            
            dist_sq = dx*dx + dy*dy + dz*dz
            mag = dt / (dist_sq * math.sqrt(dist_sq))
            
            b1.vx -= dx * b2.mass * mag
            b1.vy -= dy * b2.mass * mag
            b1.vz -= dz * b2.mass * mag
            
            b2.vx += dx * b1.mass * mag
            b2.vy += dy * b1.mass * mag
            b2.vz += dz * b1.mass * mag
            
    for i in range(n):
        b = bodies[i]
        b.x += b.vx * dt
        b.y += b.vy * dt
        b.z += b.vz * dt

def main():
    start = time.time()
    bodies = [
        Body(0, 0, 0, 100),
        Body(10, 0, 0, 10),
        Body(0, 10, 0, 10),
        Body(0, 0, 10, 10)
    ]
    
    for _ in range(1000):
        advance(bodies, 0.01)
        
    print("N-Body complete")
    elapsed = time.time() - start
    print(f"Time: {elapsed:.6f}")

if __name__ == "__main__":
    main()

File: benchmarks/reference/nbody.go
package main
import ("fmt"; "time"; "math")
type Body struct { x, y, z, mass, vx, vy, vz float64 }
func advance(bodies []Body, dt float64) {
    n := len(bodies)
    for i := 0; i < n; i++ {
        for j := i + 1; j < n; j++ {
            dx := bodies[i].x - bodies[j].x
            dy := bodies[i].y - bodies[j].y
            dz := bodies[i].z - bodies[j].z
            distSq := dx*dx + dy*dy + dz*dz
            mag := dt / (distSq * math.Sqrt(distSq))
            bodies[i].vx -= dx * bodies[j].mass * mag
            bodies[i].vy -= dy * bodies[j].mass * mag
            bodies[i].vz -= dz * bodies[j].mass * mag
            bodies[j].vx += dx * bodies[i].mass * mag
            bodies[j].vy += dy * bodies[i].mass * mag
            bodies[j].vz += dz * bodies[i].mass * mag
        }
    }
    for i := 0; i < n; i++ {
        bodies[i].x += bodies[i].vx * dt
        bodies[i].y += bodies[i].vy * dt
        bodies[i].z += bodies[i].vz * dt
    }
}
func main() {
    start := time.Now()
    bodies := []Body{
        {0, 0, 0, 100, 0, 0, 0},
        {10, 0, 0, 10, 0, 0, 0},
        {0, 10, 0, 10, 0, 0, 0},
        {0, 0, 10, 10, 0, 0, 0},
    }
    for i := 0; i < 1000; i++ { advance(bodies, 0.01) }
    fmt.Println("N-Body complete")
    fmt.Printf("Time: %.6f\n", time.Since(start).Seconds())
}
